proxy/src/control/discovery.rs

```diff
@@ -154,44 +186,89 @@ impl Discovery {
 
         Watch {
             rx,
+            metric_labels: HashMap::new(),
             bind,
         }
     }
 }
 
 // ==== impl Watch =====
 
-impl<B> Discover for Watch<B>
+impl<B> Watch<B> {
+    fn update_metadata(&mut self,
+                       addr: SocketAddr,
+                       meta: Metadata)
+                       -> Result<(), ()>
+    {
+        if let Some(store) = self.metric_labels.get_mut(&addr) {
+            store.store(meta.metric_labels)
+                .map_err(|e| {
+                    error!("update_metadata: label store error: {:?}", e);
+                })?;
+            Ok(())
+        } else {
+            // The store has already been removed, so nobody cares about
+            // the metadata change.
+            warn!(
+                "update_metadata: ignoring ChangeMetadata for {:?},
+                 the service no longer exists.",
+                addr
+            );
+            Ok(())
+        }
+    }
+}
+
+impl<B, A> Discover for Watch<B>
 where
-    B: Bind,
+    B: Bind<Request = http::Request<A>>,
 {
     type Key = SocketAddr;
     type Request = B::Request;
     type Response = B::Response;
     type Error = B::Error;
-    type Service = B::Service;
+    type Service = Labeled<B::Service>;
     type DiscoverError = ();
 
     fn poll(&mut self) -> Poll<Change<Self::Key, Self::Service>, Self::DiscoverError> {
-        let up = self.rx.poll();
-        trace!("watch: {:?}", up);
-        let update = match up {
-            Ok(Async::Ready(Some(update))) => update,
-            Ok(Async::Ready(None)) => unreachable!(),
-            Ok(Async::NotReady) => return Ok(Async::NotReady),
-            Err(_) => return Err(()),
-        };
-
-        match update {
-            Update::Insert(addr) => {
-                let service = self.bind.bind(&addr).map_err(|_| ())?;
+        loop {
+            let up = self.rx.poll();
+            trace!("watch: {:?}", up);
+            let update = match up {
```

Comment:
unrelated to your change, but this would be much simpler if it were just

```rust
let update = try_ready!(up).expect("discovery stream must be infinite");
```